File: ruby/lib/steam/packets/s2a_rules_packet.rb
# This code is free software; you can redistribute it and/or modify it under
# the terms of the new BSD License.
#
# Copyright (c) 2008-2011, Sebastian Staudt

require 'steam/packets/steam_packet'

# This class represents a S2A_RULES response sent by a game server
#
# It is used to transfer a list of server rules (a.k.a. CVARs) with their
# active values.
#
# @author Sebastian Staudt
# @see GameServer#update_rules_info
class S2A_RULES_Packet

  include SteamPacket

  # Returns the list of server rules (a.k.a. CVars) with the current values
  #
  # @return [Hash<String, String>] A list of server rules
  attr_reader :rules_hash

  # Creates a new S2A_RULES response object based on the given data
  #
  # @param [String] content_data The raw packet data sent by the server
  def initialize(content_data)
    if content_data.nil?
      raise PacketFormatError, 'Wrong formatted S2A_RULES response packet.'
    end

    super SteamPacket::S2A_RULES_HEADER, content_data

    rules_count = @content_data.short

    @rules_hash = {}

    rules_count.times do
      rule  = @content_data.cstring
      value = @content_data.cstring

      break if rule.empty?

      @rules_hash[rule] = value
    end
  end

end

File: php/lib/steam/packets/S2A_RULES_Packet.php
<?php
/**
 * This code is free software; you can redistribute it and/or modify it under
 * the terms of the new BSD License.
 *
 * Copyright (c) 2008-2009, Sebastian Staudt
 *
 * @author     Sebastian Staudt
 * @license    http://www.opensource.org/licenses/bsd-license.php New BSD License
 * @package    Steam Condenser (PHP)
 * @subpackage SteamPacket
 */

require_once "steam/packets/SteamPacket.php";

/**
 * @package Steam Condenser (PHP)
 * @subpackage SteamPacket
 */
class S2A_RULES_Packet extends SteamPacket
{
	/**
	 * @var String[]
	 */
	private $rulesArray;

	/**
	 * @param String $contentData
	 */
	public function __construct($contentData)
	{
		if(empty($contentData))
		{
			throw new Exception("Wrong formatted S2A_RULES packet.");
		}
		parent::__construct(SteamPacket::S2A_RULES_HEADER);

		$contentData = unpack("vrulesNumber/a*rulesData", $contentData);
		$tmpRulesArray = explode("\0", $contentData["rulesData"]);

		// This is a workaround for servers sending corrupt replies
		if(sizeof($tmpRulesArray) % 2) {
			array_pop($tmpRulesArray);
		}

		for($x = 0; $x < sizeof($tmpRulesArray); $x++)
		{
			$this->rulesArray[$tmpRulesArray[$x]] = $tmpRulesArray[++$x];
		}
	}

	/**
	 * @return String[]
	 */
	public function getRulesArray()
	{
		return $this->rulesArray;
	}
}
?>
